Jack/Property.fs

```diff
@@ -2,83 +2,153 @@
 
 open FSharpx.Collections
 
-type Result =
-    | Failure of List<string>
-    | Success
+type Report =
+    | Report of List<string>
+
+type Result<'a> =
+    | Failure
     | Discard
+    | Success of 'a
+
+type Property<'a> =
+    | Property of Gen<Report * Result<'a>>
+
+module Tuple =
+    let first (f : 'a -> 'c) (x : 'a, y : 'b) : 'c * 'b =
+        f x, y
 
-type Property =
-    | Property of Gen<Result>
+    let second (f : 'b -> 'c) (x : 'a, y : 'b) : 'a * 'c =
+        x, f y
 
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module Result =
-    let isFailure (x : Result) : bool =
+    let map (f : 'a -> 'b) (r : Result<'a>) : Result<'b> =
+        match r with
+        | Failure ->
+            Failure
+        | Discard ->
+            Discard
+        | Success x ->
+            Success (f x)
+
+    let filter (f : 'a -> bool) (r : Result<'a>) : Result<'a> =
+        match r with
+        | Failure ->
+            Failure
+        | Discard ->
+            Discard
+        | Success x ->
+            if f x then
+              Success x
+            else
+              Discard
+
+    let isFailure (x : Result<'a>) : bool =
         match x with
-        | Failure _ ->
+        | Failure ->
             true
-        | Success ->
-            false
         | Discard ->
             false
+        | Success _ ->
+            false
 
-    let mapFailure (f : List<string> -> List<string>) (x : Result) : Result =
-        match x with
-        | Failure msgs ->
-            Failure (f msgs)
-        | Success ->
-            Success
-        | Discard ->
-            Discard
+[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
+module Report =
+    let ofList (xs : List<string>) : Report =
+        Report xs
+
+    let toList (Report xs : Report) : List<string> =
+        xs
+
+    let empty : Report =
+        List.empty |> ofList
+
+    let map (f : List<string> -> List<string>) (xs : Report) : Report =
+        toList xs |> f |> ofList
+
+    let addFailure (msg : string) (x : Report) : Report =
+        map (List.cons msg) x
+
+    let append (xs : Report) (ys : Report) : Report =
+        toList xs @ toList ys |> ofList
 
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module Property =
+    open Tuple
```

Comment:
If we mark `Tuple` as `private`, we may then `[<AutoOpen>]` it, and then we can get rid of that `open` directive.